In [2]:
from plotly import express as px, graph_objects as go

In [3]:
import pandas as pd

In [4]:
cleaned_data = pd.read_csv("data/cleaned/terror.csv")

## Observation avant exploration

In [6]:
# nombre de colonnes et de lignes
cleaned_data.shape

(181691, 49)

On a 49 variables et 181691 observations dans le jeu de données

In [9]:
# faisons en sorte que toutes colonnes sortent lors de l'affichage
pd.options.display.max_columns = cleaned_data.shape[1]

In [10]:
# regardons à quoi ressemble les dix premières lignes
cleaned_data.head()

iyear  iday  imonth  nkill         country_txt  nwound  \
0   1970     2       7    1.0  Dominican Republic     0.0   
1   1970     0       0    0.0              Mexico     0.0   
2   1970     0       1    1.0         Philippines     0.0   
3   1970     0       1    0.0              Greece     0.0   
4   1970     0       1    0.0               Japan     0.0   

                    region_txt provstate           city  nkillus  nwoundus  \
0  Central America & Caribbean        ''  Santo Domingo      0.0       0.0   
1                North America   Federal    Mexico city      0.0       0.0   
2               Southeast Asia    Tarlac        Unknown      0.0       0.0   
3               Western Europe    Attica         Athens      0.0       0.0   
4                    East Asia   Fukouka        Fukouka      0.0       0.0   

                  region_txt.1   latitude   longitude  \
0  Central America & Caribbean  18.456792  -69.951164   
1                North America  19.371887  -99.086624   
2               Southeast Asia  15.478598  120.599741   
3               Western Europe  37.997490   23.762728   
4                    East Asia  33.580412  130.396361   

                  attacktype1_txt alternative_txt  suicide  ransompaid  \
0                   Assassination              ''        0         0.0   
1     Hostage Taking (Kidnapping)              ''        0         0.0   
2                   Assassination              ''        0         0.0   
3               Bombing/Explosion              ''        0         0.0   
4  Facility/Infrastructure Attack              ''        0         0.0   

   nhostkid hostkidoutcome_txt ransomnote  nhours  ndays  ransompaidus  \
0       0.0                 ''         ''     0.0    0.0           0.0   
1       1.0                 ''         ''     0.0    0.0           0.0   
2       0.0                 ''         ''     0.0    0.0           0.0   
3       0.0                 ''         ''     0.0    0.0           0.0   
4       0.0                 ''         ''     0.0    0.0           0.0   

   nhostkidus summary motive                               gname  \
0         0.0      ''     ''                              MANO-D   
1         0.0      ''     ''  23rd of September Communist League   
2         0.0      ''     ''                             Unknown   
3         0.0      ''     ''                             Unknown   
4         0.0      ''     ''                             Unknown   

          natlty1_txt kidhijcountry weaptype1_txt        weapsubtype1_txt  \
0  Dominican Republic            ''       Unknown                      ''   
1             Belgium        Mexico       Unknown                      ''   
2       United States            ''       Unknown                      ''   
3       United States            ''    Explosives  Unknown Explosive Type   
4       United States            ''    Incendiary                      ''   

  weaptype2_txt weapsubtype2_txt weaptype3_txt weapsubtype3_txt weaptype4_txt  \
0            ''               ''            ''               ''            ''   
1            ''               ''            ''               ''            ''   
2            ''               ''            ''               ''            ''   
3            ''               ''            ''               ''            ''   
4            ''               ''            ''               ''            ''   

  weapsubtype4_txt                targtype1_txt  \
0               ''  Private Citizens & Property   
1               ''      Government (Diplomatic)   
2               ''          Journalists & Media   
3               ''      Government (Diplomatic)   
4               ''      Government (Diplomatic)   

                                    targsubtype1_txt targtype2_txt  \
0                                     Named Civilian            ''   
1  Diplomatic Personnel (outside of embassy, cons...            ''   
2                    Radio Journalist/Staff/Facility      

Ajoutons une nouvelle colonne date qui va nous permettre de classer directement par date les observations

In [13]:
# further we must see if the column iday, imonth and iyear can be concatenated
cleaned_data["iday"].unique()

array([ 2,  0,  1,  3,  6,  8,  9, 10, 11, 12, 13, 14, 15, 19, 20, 21, 22,
       25, 26, 27, 28, 30, 31,  4,  7, 16, 17, 18, 23, 24,  5, 29],
      dtype=int64)

In [15]:
cleaned_data["imonth"].unique()

array([ 7,  0,  1,  2,  3,  4,  5,  6,  8,  9, 10, 11, 12], dtype=int64)

In [16]:
cleaned_data["iyear"].unique()

array([1970, 1971, 1972, 1973, 1974, 1975, 1976, 1977, 1978, 1979, 1980,
       1981, 1986, 1982, 1983, 1984, 1985, 1987, 1988, 1989, 1990, 1991,
       1992, 1994, 1995, 1996, 1997, 1998, 1999, 2000, 2001, 2002, 2003,
       2004, 2005, 2006, 2007, 2008, 2009, 2010, 2011, 2012, 2013, 2014,
       2015, 2016, 2017], dtype=int64)

In [17]:
cleaned_data[(cleaned_data["iday"] == 0)].index.__len__()

891

In [18]:
cleaned_data[(cleaned_data["imonth"] == 0)].index.__len__()

20

In [19]:
# Let's remove some lines by the indexes
cleaned_data.drop(index = cleaned_data[(cleaned_data["iday"] == 0)].index, axis = 0, inplace = True)

In [20]:
cleaned_data["iday"].unique()

array([ 2,  1,  3,  6,  8,  9, 10, 11, 12, 13, 14, 15, 19, 20, 21, 22, 25,
       26, 27, 28, 30, 31,  4,  7, 16, 17, 18, 23, 24,  5, 29],
      dtype=int64)

In [22]:
# create a new column named as date
cleaned_data['date'] = cleaned_data["iyear"].astype("str") + "-" + cleaned_data["imonth"].astype("str") + "-" + cleaned_data["iday"].astype("str")

In [23]:
cleaned_data["date"]

0           1970-7-2
5           1970-1-1
6           1970-1-2
7           1970-1-2
8           1970-1-2
             ...    
181686    2017-12-31
181687    2017-12-31
181688    2017-12-31
181689    2017-12-31
181690    2017-12-31
Name: date, Length: 180800, dtype: object

In [24]:
# transformation de la colonne en format date_time
cleaned_data["date"] = pd.to_datetime(cleaned_data['date'])

In [26]:
cleaned_data.sort_values('date', inplace=True)

In [27]:
cleaned_data["date"].dtype

dtype('<M8[ns]')

In [28]:
cleaned_data['date']

5        1970-01-01
10       1970-01-01
6        1970-01-02
7        1970-01-02
8        1970-01-02
            ...    
181644   2017-12-31
181645   2017-12-31
181653   2017-12-31
181676   2017-12-31
181690   2017-12-31
Name: date, Length: 180800, dtype: datetime64[ns]

In [30]:
cleaned_data.to_csv('data/cleaned/terror.csv', index=False)

# Visualiser le nombre de morts causés par des attaques terroristes par pays et par année

variables : nkill, country_txt, iyear

Récupération des données par pays et années et aggrégeons par morts

In [32]:
# créons un dataframe qui prend la colonne date comme index
cleaned_data_2 = cleaned_data.set_index('date')

In [37]:
cleaned_data_2.head()

iyear  iday  imonth  nkill    country_txt  nwound     region_txt  \
date                                                                           
1970-01-01   1970     1       1    0.0  United States     0.0  North America   
1970-01-01   1970     1       1    0.0  United States     0.0  North America   
1970-01-02   1970     2       1    0.0        Uruguay     0.0  South America   
1970-01-02   1970     2       1    0.0  United States     0.0  North America   
1970-01-02   1970     2       1    0.0  United States     0.0  North America   

             provstate        city  nkillus  nwoundus   region_txt.1  \
date                                                                   
1970-01-01    Illinois       Cairo      0.0       0.0  North America   
1970-01-01   Wisconsin     Baraboo      0.0       0.0  North America   
1970-01-02  Montevideo  Montevideo      0.0       0.0  South America   
1970-01-02  California     Oakland      0.0       0.0  North America   
1970-01-02   Wisconsin     Madison      0.0       0.0  North America   

             latitude   longitude                 attacktype1_txt  \
date                                                                
1970-01-01  37.005105  -89.176269                   Armed Assault   
1970-01-01  43.468500  -89.744299               Bombing/Explosion   
1970-01-02 -34.891151  -56.187214                   Assassination   
1970-01-02  37.791927 -122.225906               Bombing/Explosion   
1970-01-02  43.076592  -89.412488  Facility/Infrastructure Attack   

                       alternative_txt  suicide  ransompaid  nhostkid  \
date                                                                    
1970-01-01                          ''        0         0.0       0.0   
1970-01-01  Insurgency/Guerilla Action        0         0.0       0.0   
1970-01-02                          ''        0         0.0       0.0   
1970-01-02            Other Crime Type        0         0.0       0.0   
1970-01-02                          ''        0         0.0       0.0   

           hostkidoutcome_txt ransomnote  nhours  ndays  ransompaidus  \
date                                                                    
1970-01-01                 ''         ''     0.0    0.0           0.0   
1970-01-01                 ''         ''     0.0    0.0           0.0   
1970-01-02                 ''         ''     0.0    0.0           0.0   
1970-01-02                 ''         ''     0.0    0.0           0.0   
1970-01-02                 ''         ''     0.0    0.0           0.0   

            nhostkidus                                            summary  \
date                                                                        
1970-01-01         0.0  1/1/1970: Unknown African American assailants ...   
1970-01-01         0.0                                                 ''   
1970-01-02         0.0                                                 ''   
1970-01-02         0.0  1/2/1970: Unknown perpetrators detonated explo...   
1970-01-02         0.0  1/2/1970: Karl Armstrong, a member of the New ...   

                                                    motive  \
date                                                         
1970-01-01  To protest the Cairo Illinois Police Deparment   
1970-01-01                                              ''   
1970-01-02                                              ''   
1970-01-02                                              ''   
1970-01-02     To protest the War in Vietnam and the draft   

                                      gname    natlty1_txt kidhijcountry  \
date                                                                       
1970-01-01               Black Nationalists  United States            ''   
1970-01-01  Weather Underground, Weathermen  United States            ''   
1970-01-02              Tupamaros (Uruguay)        Uruguay            ''   
1970-01-02                          Unknown  United States            ''   
1970-01

In [38]:
df_pays_morts = cleaned_data_2.groupby(['country_txt', 'iyear'], as_index=False).sum('nkill')

In [39]:
df_pays_morts.head()

country_txt  iyear  iday  imonth  nkill  nwound  nkillus  nwoundus  \
0  Afghanistan   1973     1       5    0.0     1.0      0.0       0.0   
1  Afghanistan   1979    50      19   53.0     1.0      1.0       0.0   
2  Afghanistan   1987    31       5    0.0     2.0      0.0       0.0   
3  Afghanistan   1988   148      75  128.0   106.0      0.0       0.0   
4  Afghanistan   1989   168      61   10.0    84.0      0.0       0.0   

     latitude   longitude  suicide  ransompaid  nhostkid  nhours  ndays  \
0   34.516895   69.147011        0         0.0       0.0     0.0    0.0   
1  102.406239  199.759655        0         0.0       1.0     0.0    0.0   
2    0.000000    0.000000        0         0.0       0.0     0.0    0.0   
3  380.645164  759.021904        0         0.0       0.0     0.0    0.0   
4  341.883293  674.938996        0         0.0       0.0     0.0    0.0   

   ransompaidus  nhostkidus  propextent  propvalue  
0           0.0         0.0         0.0        0.0  
1           0.0         1.0         3.0    30000.0  
2           0.0         0.0         0.0        0.0  
3           0.0         0.0        17.0  2135000.0  
4           0.0         0.0         3.0   100000.0

In [40]:
px.line(df_pays_morts, x = "iyear", y = "nkill", hue = "country_txt")

TypeError: line() got an unexpected keyword argument 'hue'